node/src/frost_sign.rs

```diff
@@ -0,0 +1,576 @@
+use std::collections::btree_map::Iter;
+use crate::primitives::ParticipantId;
+use aes_gcm::aead::rand_core::{CryptoRng, RngCore};
+use anyhow::Context;
+use borsh::{BorshDeserialize, BorshSerialize};
+use cait_sith::protocol::{Action, MessageData, Participant, Protocol, ProtocolError};
+use std::collections::BTreeMap;
+
+/// This implementation is valid in the context where we can choose a Leader (Coordinator)
+/// node for a signature generation. That means only Coordinator ends up with a generated signature.
+/// It comes with a benefit of O(n) messages in a network.
+///
+/// If one wants to get rid of a such distinction, it will make the number of messages in the system O(n^2).
+///
+/// TODO: Can we precompute the first round in advance (just like presigns in ECDSA)?
+///     The answer depends on whether we can do KDF on SigningPackage/SigningShare.
+
+#[derive(Debug)]
+pub enum FrostOutput {
+    Coordinator(frost_ed25519::Signature),
+    Participant,
+}
+
+pub fn frost_sign<RNG: CryptoRng + RngCore + 'static>(
```

Comment:
Here and below: could you sprinkle some documentation so the reader of this code can more easily follow it? Mostly I'm looking for documentation for important methods whose behavior is not immediately clear, structures fields whose meaning is not immediately clear, etc.